internal/server/config.go

```diff
@@ -9,65 +9,116 @@ import (
 	"strings"
 	"time"
 
-	"github.com/go-playground/validator/v10"
 	"github.com/infrahq/secrets"
 	"golang.org/x/crypto/bcrypt"
 	"gorm.io/gorm"
 
+	"github.com/infrahq/infra/api"
 	"github.com/infrahq/infra/internal"
 	"github.com/infrahq/infra/internal/logging"
 	"github.com/infrahq/infra/internal/server/data"
 	"github.com/infrahq/infra/internal/server/models"
 	"github.com/infrahq/infra/internal/server/providers"
+	"github.com/infrahq/infra/internal/validate"
 	"github.com/infrahq/infra/uid"
 )
 
 type Provider struct {
-	Name         string `validate:"required"`
-	URL          string `validate:"required"`
-	ClientID     string `validate:"required"`
-	ClientSecret string `validate:"required"`
+	Name         string
+	URL          string
+	ClientID     string
+	ClientSecret string
 	Kind         string
 	AuthURL      string
 	Scopes       []string
 }
 
+func (p Provider) ValidationRules() []validate.ValidationRule {
+	return []validate.ValidationRule{
+		api.ValidateName(p.Name),
+		validate.Required("name", p.Name),
+		validate.Required("url", p.URL),
+		validate.Required("clientID", p.ClientID),
+		validate.Required("clientSecret", p.ClientSecret),
+	}
+}
+
 type Grant struct {
-	User     string `validate:"excluded_with=Group,excluded_with=Machine"`
-	Group    string `validate:"excluded_with=User,excluded_with=Machine"`
-	Machine  string `validate:"excluded_with=User,excluded_with=Group"` // deprecated
-	Resource string `validate:"required"`
+	User     string
+	Group    string
+	Machine  string // deprecated
```

Comment:
Can probably remove `machine` now

Reply:
I was wondering about that. Going to leave it for a follow up in case there are other implications to removing it.